### PR TITLE
Feature/77 allow work labels on user mode

### PR DESCRIPTION
**Pull request by pcantera**
# Description

# Related Issue

# Motivation and Context

# How Has This Been Tested?

# Screenshots (if appropriate)

# Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

# Checklist

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: api/types/labels.go [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+type Label struct {
+	ID           string `json:"id" header:"ID"`
+	Name         string `json:"name" header:"NAME"`
+	ResourceType string `json:"resource_type" header:"RESOURCE_TYPE"`
+	Namespace    string `json:"namespace" header:"NAMESPACE" show:"nolist"`
+	Value        string `json:"value" header:"VALUE" show:"nolist"`
+}
+
+type LabeledResources struct {
```
Comment (pbanos) [inline api/types/labels.go]: A better name for this might be `LabeledResource`

File: blueprint/scripts/subcommands.go [hunk 2]
```diff
@@ -84,5 +94,47 @@ func SubCommands() []cli.Command {
 				},
 			},
 		},
+		{
+			Name:   "add-label",
+			Usage:  "This action assign a single label from a single labelable resource",
```
Comment (pbanos) [inline blueprint/scripts/subcommands.go]: The usage message should read as `This action assigns [...]`

File: blueprint/scripts/subcommands.go [hunk 3]
```diff
@@ -84,5 +94,47 @@ func SubCommands() []cli.Command {
...
+					Name:   "resource_type",
+					Usage:  "Resource Type",
+					Value:  "script",
+					Hidden: true,
+				},
+			},
+		},
+		{
+			Name:   "remove-label",
+			Usage:  "This action de-assign a single label from a single labelable resource",
```
Comment (pbanos) [inline blueprint/scripts/subcommands.go]: I would use as verb here either remove or unassign, and it also needs an s: `This action unassigns [...]`

File: cmd/labels_cmd.go [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+	}
+	if err = formatter.PrintItem(*label); err != nil {
+		formatter.PrintFatal("Couldn't print/format result", err)
+	}
+	return nil
+}
+
+// LabelFiltering subcommand function receives an interface representing a collection of labelable resources (Server, Template, ...)
+// Evaluates the matching of assigned labels with the labels requested for filtering.
+func LabelFiltering(c *cli.Context, items interface{}) (*[]interface{}, error) {
```
Comment (pbanos) [inline cmd/labels_cmd.go]: The code for LabelFiltering here and LabelAssignNamesForIDs below works as expected but relays heavily on reflection and exposes too broad an interface using the generic interface{} for input and output types. I think an alternative that uses type composition may be possible and more idiomatic.

Labelable resources would be composed with a LabelableField struct containing the LabelIDs and Labels slices of strings. 
```Go
type LabelableField struct {
  LabelIDs              []string         `json:"label_ids" header:"LABEL_IDS" show:"nolist,noshow"`
  Labels                 []string         `json:"labels" header:"LABELS"`
}

type FirewallProfile struct {
ID string
[...]
LabelableFields
}
```
We would also have a Labelable interface with a set of methods all labelable resources should implement. I'm thinking of a label ID filter and a label IDs and a label names assigning methods but there may be other methods we have to include in it. 
```Go
type Labelable interface {
 FilterByLabelIDs(labelIDs []string) bool
 AssignLabelIDs(labelIDs []string)
 FillInLabelNames(labelNamesByID map[string]string)
}
```
By making LabelableFields implement Labelable, we ensure the different labelable resources (Server, FirewallProfile, etc) implement Labelable:
```Go
func (lf *LabelableFields) FilterByLabelIDs(labelIDs []string) bool {
  for _, lid := range labelIDs {
    var labelIDFound bool
    for _, resourceLabelID := range lf.LabelIDs {
       if lID == resourceLabelID {
         labelIDFound = true
         break;
       }
    }
    if !labelIDFound {
      return false
   }
  }
  return true
}

[...]
```

That way, LabelFiltering could take and return slices of Labelable as input and output. We would still need to convert slices of specific labelable resources (slices of servers, slices of firewall profiles, etc) into slices of labelables, but we would benefit from static typing and clearer code I think...

Comment (pcantera) [reply cmd/labels_cmd.go]: Done

File: utils/utils.go [hunk 5]
```diff
@@ -176,3 +177,44 @@ func RandomString(strlen int) string {
...
+		if x == n {
+			return true
+		}
+	}
+	return false
+}
+
+// Subset returns true if the first slice is completely contained in the second slice.
+// There must be at least the same number of duplicate values in second as there are in first.
+func Subset(s1, s2 []string) bool {
```
Comment (pbanos) [inline utils/utils.go]: I think this method is not used anywhere in the codebase

Comment (pcantera) [reply utils/utils.go]: yes, there are a lot of them... My idea is to remove all of them in close future, refactoring and normalizing focused in these aspects

**Pull request merged**